### PR TITLE
Add doc about JsonSerializer collection support in 3.1

### PR DESCRIPTION
**Pull request by layomia**

### CODE REVIEW

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+| ImmutableStack<T> | Supported | Supported |
+| IImmutableDictionary<string, TValue> | Supported | Supported |
+| IImmutableList<T> | Supported | Supported |
+| IImmutableQueue<T> | Supported | Supported |
+| IImmutableSet<T> | Supported | Supported |
+| IImmutableStack<T> | Supported | Supported |
+
+### System.Collections.Specialized
+
+#### Deserialization heuristics
```
Comment (tdykstra) [inline src/libraries/System.Text.Json/docs/collection_support.md]: What does this mean?

**Review by tdykstra — APPROVED** [review]
Looks great!  I had one question.

**Comment by jkotas** [comment]
Should user facing documentation like this rather go to the official docs at https://docs.microsoft.com/en-us/dotnet/api/system.text.json?view=netcore-3.1 ?

cc @carlossanlop @danmosemsft

File: src/libraries/System.Text.Json/docs/collection_support.md (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/docs/collection_support.md]: I had notes earlier about how the deserialization is done for each class of types. This is not relevant is this doc so I removed them (and forgot this one). I just pushed a commit to remove this line.

**Comment by layomia** [comment]
This is a placeholder for a larger doc that @tdykstra plans to write about supported types in System.Text.Json: https://github.com/dotnet/docs/issues/15807.

The plan is to link to this doc from the [official docs](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-how-to) in the mean time.

**Comment by jkotas** [comment]
What would be wrong with putting this placeholder doc into the docs repo, and iterate on it there?

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+collection not in the BCL, (i.e. outside System.Collections[.[.*]]), is considered
+a custom collection. This includes user-defined types and ASP.NET defined types,
+e.g. those in
+[Microsoft.Extensions.Primitives](https://docs.microsoft.com/dotnet/api/microsoft.extensions.primitives?view=dotnet-plat-ext-3.1).
+
+A custom collection is supported for deserialization if it fulfils the following:
+
+1. Is not an interface or abstract.
+2. Has a parameter-less constructor
+3. Implements one or more of IList, IList<T>, ICollection<T>, Stack<T>, Queue<T>,
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]:
```suggestion
3. Implements one or more of `IList`, `IList<T>`, `ICollection<T>`, `Stack<T>`, `Queue<T>`,
```

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+   Stack, Queue, IDictionary, IDictionary<string, TValue>
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]:
```suggestion
   `Stack`, `Queue`, `IDictionary`, `IDictionary<string, TValue>`
```

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+4. The element type is supported in System.Text.Json
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]: Here and elsewhere.
```suggestion
4. The element type is supported in `System.Text.Json`
```

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
+# JsonSerializer collection support in 3.1.x
+
+All collections (any type that derives from IEnumerable) are supported for
+serialization with JsonSerializer; we simply call the .GetEnumerator() method
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]:
```suggestion
serialization with `JsonSerializer`; we simply call the `.GetEnumerator()` method
```

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
+# JsonSerializer collection support in 3.1.x
+
+All collections (any type that derives from IEnumerable) are supported for
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]: Consider making all references to types/collections, links to the API docs for those types.
```suggestion
All collections (any type that derives from `IEnumerable`) are supported for
```

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+A custom collection is supported for deserialization if it fulfils the following:
+
+1. Is not an interface or abstract.
+2. Has a parameter-less constructor
+3. Implements one or more of IList, IList<T>, ICollection<T>, Stack<T>, Queue<T>,
+   Stack, Queue, IDictionary, IDictionary<string, TValue>
+4. The element type is supported in System.Text.Json
+
+All custom collections are supported for serialization (everything that derives from IEnumerable),
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]:
```suggestion
All custom collections are supported for serialization (everything that derives from `IEnumerable`),
```

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+as long as their element types are supported.
+
+There are known issues with some custom collections where we don't offer rountrippable support.
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]:
```suggestion
There are known issues with some custom collections where we don't offer round-trippable support.
```

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+| ObservableCollection<T> | Supported | Supported |
+| KeyedCollection<string, TValue> | Supported | Not supported |
+| ReadOnlyCollection<T> | Supported | Not supported |
+| ReadOnlyObservableCollection<T> | Supported | Not supported |
+| ReadOnlyDictionary<string, TValue> | Supported | Not supported |
+
+## Custom collections
+
+For the purpose of serialization and deserialization in System.Text.Json, any
+collection not in the BCL, (i.e. outside System.Collections[.[.*]]), is considered
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]: Why is everything after the regex italicized? Would back-ticks fix it?

```suggestion
collection not in the BCL, (i.e. outside `System.Collections[.[.*]]`), is considered
```

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+| Dictionary<string, TValue> | Supported | Supported |
+| HashSet<T> | Supported | Supported |
+| LinkedList<T> | Supported | Supported |
+| LinkedListNode<T> | Supported | Not supported |
+| List<T> | Supported | Supported |
+| Queue<T> | Supported | Supported |
+| SortedDictionary<string,TValue> | Supported | Supported |
+| SortedList<string,TValue> | Supported | Supported |
+| SortedSet<T> | Supported | Supported |
+| Stack<T> | Supported | Supported |
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]: Should we put an asterisk next to those where the behavior might be unexpected (so we can add notes about them below the table)? For example, serializing `Stack<T>` types.

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+All custom collections are supported for serialization (everything that derives from IEnumerable),
+as long as their element types are supported.
+
+There are known issues with some custom collections where we don't offer rountrippable support.
+These include:
+
+- ExpandoObject: https://github.com/dotnet/corefx/issues/38007
+- DynamicObject: https://github.com/dotnet/corefx/issues/41105 
+- DataTable: https://github.com/dotnet/corefx/issues/38712 
+- Microsoft.AspNetCore.Http.FormFile: https://github.com/dotnet/corefx/issues/41401
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]: What about others:
https://github.com/dotnet/corefx/issues/42084
https://github.com/dotnet/corefx/issues/41105
Maybe? https://github.com/dotnet/corefx/issues/42399
Maybe? https://github.com/dotnet/corefx/issues/41399

**Comment by tdykstra** [comment]
@jkotas 

>What would be wrong with putting this placeholder doc into the docs repo, and iterate on it there?

This is only part of the planned supported types doc. It's a source for that doc, and format and content when incorporated into a publishable doc will be different.

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+| [`ImmutableDictionary<string, TValue>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutabledictionary-2?view=netcore-3.1) | Supported | Supported |
+| [`ImmutableHashSet<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablehashset-1?view=netcore-3.1) | Supported | Supported |
+| [`IImmutableList<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.iimmutablelist-1?view=netcore-3.1) | Supported | Supported |
+| [`ImmutableQueue<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablequeue-1?view=netcore-3.1) | Supported | Supported |
+| [`ImmutableSortedDictionary<string, TValue>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablesorteddictionary-2?view=netcore-3.1) | Supported | Supported |
+| [`ImmutableSortedSet<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablesortedset-1?view=netcore-3.1) | Supported | Supported |
+| [`ImmutableStack<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablestack-1?view=netcore-3.1) | Supported | Supported |
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]: Add `*` to these other Stack types

File: src/libraries/System.Text.Json/docs/collection_support.md [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+- Support for [`DynamicObject`](https://docs.microsoft.com/dotnet/api/system.dynamic.dynamicobject?view=netcore-3.1): https://github.com/dotnet/corefx/issues/41105
+- Support for [`DataTable`](https://docs.microsoft.com/dotnet/api/system.data.datatable?view=netcore-3.1): https://github.com/dotnet/corefx/issues/38712
+- Support for [`FormFile`](https://docs.microsoft.com/dotnet/api/microsoft.aspnetcore.http.formfile?view=aspnetcore-3.1): https://github.com/dotnet/corefx/issues/41401
+- Support for [`IFormCollection`](https://docs.microsoft.com/dotnet/api/microsoft.aspnetcore.http.iformcollection?view=aspnetcore-3.1)
+- Assigning `null` to value-type collections like [`ImmutableArray<T>`](https://docs.microsoft.com/dotnet/api/system.collections.immutable.immutablearray-1?view=netcore-3.1): https://github.com/dotnet/corefx/issues/42399
+
+For more information, see the [open issues in System.Text.Json](https://github.com/dotnet/runtime/issues?q=is%3Aopen+is%3Aissue+label%3Aarea-System.Text.Json).
+
+\* All instances of custom collections that implement stack types are reversed on serialization,
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/docs/collection_support.md]: Was repeating this intentional?

**Comment by ahsonkhan** [comment]
> What would be wrong with putting this placeholder doc into the docs repo, and iterate on it there?

I would be fine with that. We don't _have_ to merge this doc here in dotnet/runtime, and can wait for it to be incorporated into the "supported types" page in the official docs.

Maybe, the outcome of this PR is a posting this content as a comment into the https://github.com/dotnet/docs/issues/15807, as reference material.

**Comment by tdykstra** [comment]
>Maybe, the outcome of this PR is a posting this content as a comment into the dotnet/docs#15807, as reference material.

I was going to link to it in the description box

File: src/libraries/System.Text.Json/docs/collection_support.md (hunk 13 above)
Comment (layomia) [reply src/libraries/System.Text.Json/docs/collection_support.md]: Yes, there was a distinction between built-in types and custom types. I've consolidated these notes at the bottom.

**Comment by layomia** [comment]
Thanks all for the review. Per the conversation above, I'm going to close this PR. The contents can be applied to https://github.com/dotnet/docs/issues/15807 as needed.

**Comment by tdykstra** [comment]
In that case I'll add it as a comment in https://github.com/dotnet/docs/issues/15807 as suggested by @ahsonkhan

**Pull request closed without merging**